crypto/aes/build.info

```diff
@@ -2,15 +2,15 @@ LIBS=../../libcrypto
 
 $AESASM=aes_core.c aes_cbc.c
 IF[{- !$disabled{asm} -}]
-  $AESASM_x86=aes-586.s
-  $AESDEF_x86=AES_ASM
+  $AESASM_x86=aes_core.c aes_cbc.c
+  $AESDEF_x86=
```

Comment:
I know this seems counterintuitive, but `AES_ASM` should still be defined

Reply:
I tried to do that, but aes_core.c will stop to function if AES_ASM is defined:
```
diff --git a/crypto/aes/build.info b/crypto/aes/build.info
index fb67a7b..2020268 100644
--- a/crypto/aes/build.info
+++ b/crypto/aes/build.info
@@ -5,12 +5,12 @@ IF[{- !$disabled{asm} -}]
   $AESASM_x86=aes_core.c aes_cbc.c
   $AESDEF_x86=
   $AESASM_x86_sse2=vpaes-x86.s aesni-x86.s
-  $AESDEF_x86_sse2=VPAES_ASM
+  $AESDEF_x86_sse2=AES_ASM VPAES_ASM
 
   $AESASM_x86_64=\
         aes_core.c aes_cbc.c vpaes-x86_64.s aesni-x86_64.s \
         aesni-sha1-x86_64.s aesni-sha256-x86_64.s aesni-mb-x86_64.s
-  $AESDEF_x86_64=VPAES_ASM
+  $AESDEF_x86_64=AES_ASM VPAES_ASM
 
   $AESASM_ia64=aes_core.c aes_cbc.c aes-ia64.s
   $AESDEF_ia64=AES_ASM
diff --git a/crypto/include/internal/ciphermode_platform.h b/crypto/include/internal/ciphermode_platform.h
index e55654f..6001428 100644
--- a/crypto/include/internal/ciphermode_platform.h
+++ b/crypto/include/internal/ciphermode_platform.h
@@ -94,7 +94,7 @@ void AES_xts_decrypt(const unsigned char *inp, unsigned char *out, size_t len,
 #  endif
 # endif /* OPENSSL_CPUID_OBJ */
 
-# if     !defined(OPENSSL_NO_ASM) &&                     (  \
+# if     defined(AES_ASM) &&                             (  \
          ((defined(__i386)       || defined(__i386__)    || \
            defined(_M_IX86)) && defined(OPENSSL_IA32_SSE2))|| \
          defined(__x86_64)       || defined(__x86_64__)  || \
```
.-> 
```
gcc  -Iinclude  -pthread -m64 -Wa,--noexecstack -Wall -O3 -DNDEBUG -DOPENSSL_IA32_SSE2  -MMD -MF test/buildtest_c_asn1t-bin-buildtest_asn1t.d.tmp -MT test/buildtest_c_asn1t-bin-buildtest_asn1t.o -c -o test/buildtest_c_asn1t-bin-buildtest_asn1t.o test/buildtest_asn1t.c
crypto/aes/fips-dso-aes_cbc.o: In function `AES_cbc_encrypt':
aes_cbc.c:(.text+0x8): undefined reference to `AES_decrypt'
aes_cbc.c:(.text+0x1b): undefined reference to `AES_encrypt'
crypto/aes/fips-dso-aes_ecb.o: In function `AES_ecb_encrypt':
aes_ecb.c:(.text+0x6): undefined reference to `AES_decrypt'
aes_ecb.c:(.text+0x11): undefined reference to `AES_encrypt'
providers/common/ciphers/fips-dso-cipher_aes_hw.o: In function `cipher_hw_aes_initkey':
cipher_aes_hw.c:(.text+0x3d): undefined reference to `AES_encrypt'
cipher_aes_hw.c:(.text+0x120): undefined reference to `AES_decrypt'
providers/common/ciphers/fips-dso-cipher_ccm_hw.o: In function `ccm_generic_aes_initkey':
cipher_ccm_hw.c:(.text+0x252): undefined reference to `AES_encrypt'
providers/common/ciphers/fips-dso-cipher_gcm_hw.o: In function `generic_aes_gcm_initkey':
cipher_gcm_hw.c:(.text+0x299): undefined reference to `AES_encrypt'
collect2: error: ld returned 1 exit status
```

Reply:
Ah!  Figured it out by looking at how this is done in 1.1.1.  From Configure:

``` perl
        push @{$config{lib_defines}}, "AES_ASM" if ($target{aes_asm_src} =~ m/\baes-/);;
```

So you're right, `AES_ASM` shouldn't be defined here since only `aes_core.c` and `aes_cbc.c` are used.